### PR TITLE
conf: watch for changes in config override files

### PR DESCRIPTION
**Pull request by keegancsmith**
The main reason for doing this is for allowing frontend to update the
configuration when kubernetes updates configuration files (from the
configmap mounts). Before this change we required restarting frontend
when pushing new configuration changes.

### CODE REVIEW

**Comment by pecigonzalo** [comment]
While I really like the idea, I would not enable this functionality as it would mean that changes will not be rolled to instances and rather be deployed all at once.
To support this change, we should have a way to roll this change to avoid any failures.

**Comment by keegancsmith** [comment]
> While I really like the idea, I would not enable this functionality as it would mean that changes will not be rolled to instances and rather be deployed all at once.

That is how it works today, except instead of being related to the value of the configmap it is related to the last time a sourcegraph-frontend instance started. IE the way config file overrides work is by updating the state in postgres, not by updating in process state. There have been a few times where people at sourcegraph have updated the configuration and then wondered why it hadn't deployed yet (new homepage rollout and some marketing stuff as an example).

If we wanted a way to incrementally do config rollout that would be a very different change.

**Comment by pecigonzalo** [comment]
Ok, I was not aware that was the case. So `frontend reads the file -> writes to the DB -> services query the frontend -> frontend reads from the DB`?
Do the frontends themselves also read the DB for their own config?

Thanks for clarifying this.

File: cmd/frontend/internal/cli/config.go [hunk 1]
```diff
@@ -203,9 +206,81 @@ func handleConfigOverrides() error {
 		}
 	}
 
+	// Kick off a background fsnotify watcher of the config files.
+	go configOverridesWatchOnce.Do(func() {
+		events, err := watchPaths(ctx, overrideSiteConfig, overrideExtSvcConfig, overrideGlobalSettings)
+		if err != nil {
+			log15.Error("failed to watch config override files", "error", err)
```
Comment (uwedeportivo) [inline cmd/frontend/internal/cli/config.go:213]: i agree that this isn't fatal and it should continue with its business. i would add an expvar that shows the state of the watcher (watching or failed). hunting in the logs for this state is more work given our verbose logs :-)

**Review by uwedeportivo — APPROVED** [review]

File: cmd/frontend/internal/cli/config.go (hunk 1 above)
Comment (keegancsmith) [reply cmd/frontend/internal/cli/config.go:213]: expvar or prometheus metric? :)

Comment (uwedeportivo) [reply cmd/frontend/internal/cli/config.go:213]: either one :-). whichever is easier. consumption of it is mostly the same. with prometheus we could alert i guess.

**Comment by slimsag** [comment]
@keegancsmith it seems this does not work in practice for some reason. Perhaps updating a configmap does not actually cause the file to change on disk in k8s? See https://sourcegraph.slack.com/archives/CMBA8F926/p1600734032009700

**Pull request merged**